tz.cpp

```diff
@@ -3087,121 +3124,145 @@ locate_native_zone(const std::string& native_tz_name)
-            in.read(buffer.data(), 4);
-            char_count = detail::interpret32(&buffer[0]);
-        }
-        if(leaps)
-        {
-            if(leap_count == 0)
-                throw std::runtime_error{"No leaps in "+name_};
-            leaps->clear();
-            leaps->reserve(leap_count);
-            in.ignore(
-                (time_size+1) * time_count +
-                6 * type_count +
-                char_count
-               );
-            size_t buffer_size = time_size + 4;
-            for(std::uint32_t leap_last = 0; leap_last < leap_count;) {
-                if(!in.read(buffer.data(), buffer_size))break;
-                std::uint32_t count = detail::interpret32(&buffer[time_size]);
-                if(++leap_last != count) {
-                    throw std::runtime_error{"invalid tzfile"};
+    }
+    p.transitions_.reserve(time_count);
+    for(std::uint32_t i = 0; i < time_count; ++i) {
+        in.read(buffer.data(), time_size);
+        std::chrono::seconds s{detail::interpret_time(buffer.data(), time_size)};
+        p.transitions_.emplace_back(sys_seconds{s});
+    }
+    std::vector<std::uint8_t> transition_indexes(time_count, 0);
+    in.read(reinterpret_cast<char*>(transition_indexes.data()), transition_indexes.size());
+    p.local_infos_.reserve(type_count);
+    std::vector<std::uint8_t> abbreviation_indexes;
+    abbreviation_indexes.reserve(type_count);
+    for(; type_count > 0; --type_count) {
+        in.read(buffer.data(), 6);
+        std::chrono::seconds s{detail::interpret_time(buffer.data(), 4)};
+        p.local_infos_.emplace_back(s, buffer[4] != 0, std::string{});
+        abbreviation_indexes.push_back(buffer[5]);
+        if(abbreviation_indexes.back() >= char_count) {
+            throw std::runtime_error{"invalid tzfile: abbreviation index "+std::to_string(abbreviation_indexes.back())+" >= "+std::to_string(char_count)};
+        }
+    }
+    for(size_t i = 0; i < transition_indexes.size(); ++i) {
+        p.transitions_[i].info = &p.local_infos_.at(transition_indexes[i]);
+    }
+    std::unique_ptr<char[]> abbreviations(new char[char_count+1]);
+    if(char_count > 0) {
+        in.read(abbreviations.get(), char_count);
+        abbreviations[char_count] = '\0';
+        for(size_t i = 0; i < abbreviation_indexes.size(); ++i) {
+            p.local_infos_[i].abbreviation = &abbreviations[abbreviation_indexes[i]];
+        }
+    }
+    std::vector<leap> leaps_;
+    leaps_.reserve(leap_count);
+    size_t buffer_size = time_size + 4;
+    for(std::uint32_t leap_last = 0; leap_last < leap_count;) {
+        if(!in.read(buffer.data(), buffer_size))break;
+        std::uint32_t count = detail::interpret32(&buffer[time_size]);
+        if(++leap_last != count) {
+            throw std::runtime_error{"invalid tzfile"};
+        }
+        leaps_.emplace_back(sys_seconds{std::chrono::seconds{detail::interpret_time(buffer.data(), time_size)}});
+    }
+    if(!leaps_.empty()) {
```

Comment:
Is `!leaps_.empty()` ever true when reading files not under `right`?  I'm wondering if this part can just be ignored.